File: android/app/src/main/java/com/forrestlasiter/umbra/net/dns/TelemetryBlocklist.kt
package com.forrestlasiter.umbra.net.dns

import com.forrestlasiter.umbra.core.CoreSpec

/**
 * Domain matcher for the DNS sinkhole. Domains come from the shared core spec
 * (the same lists the Linux /etc/hosts sinkhole uses), so the phone and the
 * laptop block exactly the same telemetry.
 */
class TelemetryBlocklist(domains: Set<String>) {

    private val blocked: Set<String> = domains.map { it.lowercase().trimEnd('.') }.toSet()

    val size: Int get() = blocked.size

    /** True if [qName] is a blocked domain or a subdomain of one. */
    fun isBlocked(qName: String): Boolean {
        val name = qName.lowercase().trimEnd('.')
        if (name in blocked) return true
        // walk parent domains: a.b.example.com -> b.example.com -> example.com
        var idx = name.indexOf('.')
        while (idx != -1) {
            if (name.substring(idx + 1) in blocked) return true
            idx = name.indexOf('.', idx + 1)
        }
        return false
    }

    companion object {
        fun forProfile(spec: CoreSpec, profile: String): TelemetryBlocklist =
            TelemetryBlocklist(spec.telemetryDomains(profile))
    }
}

File: ios/UmbraCore/Sources/UmbraCore/Net/TelemetryBlocklist.swift
import Foundation

/// Domain matcher for the DNS sinkhole, built from the shared core spec — the
/// same lists the Linux /etc/hosts sinkhole and the Android app use.
public struct TelemetryBlocklist {
    private let blocked: Set<String>
    public var count: Int { blocked.count }

    public init(_ domains: Set<String>) {
        blocked = Set(domains.map { $0.lowercased().trimmingTrailingDots() })
    }

    public init(spec: CoreSpec, profile: String) {
        self.init(spec.telemetryDomains(profile))
    }

    /// True if `qName` is a blocked domain or a subdomain of one.
    public func isBlocked(_ qName: String) -> Bool {
        let name = qName.lowercased().trimmingTrailingDots()
        if blocked.contains(name) { return true }
        var rest = name
        while let dot = rest.firstIndex(of: ".") {
            rest = String(rest[rest.index(after: dot)...])
            if blocked.contains(rest) { return true }
        }
        return false
    }
}

private extension String {
    func trimmingTrailingDots() -> String {
        var s = self
        while s.hasSuffix(".") { s.removeLast() }
        return s
    }
}
